ereport/types/src/lib.rs

```diff
@@ -0,0 +1,145 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Core types for representing ereports.
+
+use core::fmt;
+pub use omicron_uuid_kinds::EreporterGenerationUuid;
+use schemars::JsonSchema;
+use serde::Deserialize;
+use serde::Serialize;
+use uuid::Uuid;
+
+/// An ereport message.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize, JsonSchema)]
+pub struct Ereport {
+    /// The ENA of the ereport.
+    pub ena: Ena,
+    /// The body of the ereport.
+    pub report: ReportKind,
+}
+
+/// An Error Numeric Association (ENA)
+#[derive(
+    Copy,
+    Clone,
+    Eq,
+    PartialEq,
+    Ord,
+    PartialOrd,
+    Serialize,
+    Deserialize,
+    JsonSchema,
+)]
+#[repr(transparent)]
+pub struct Ena(pub u64);
+
+impl fmt::Display for Ena {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:#x}", self.0)
+    }
+}
+
+impl fmt::Debug for Ena {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Ena({:#x})", self.0)
+    }
+}
+
+impl fmt::UpperHex for Ena {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::UpperHex::fmt(&self.0, f)
+    }
+}
+
+impl fmt::LowerHex for Ena {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::LowerHex::fmt(&self.0, f)
+    }
+}
+/// Uniquely identifies the entity that generated an ereport.
+#[derive(
+    Copy, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, JsonSchema,
+)]
+pub struct Reporter {
+    pub reporter_id: Uuid,
+    pub gen_id: EreporterGenerationUuid,
+}
+
+/// The body of an ereport: either an event is reported, or a loss report.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum ReportKind {
+    /// An ereport.
+    Event(Event),
+    /// Ereports were lost, or may have been lost.
+    Loss {
+        // The number of ereports that were discarded, if it is known.
+        ///
+        /// If ereports are dropped because a buffer has reached its capacity,
+        /// the reporter is strongly encouraged to attempt to count the number
+        /// of ereports lost. In other cases, such as a reporter crashing and
+        /// restarting, the reporter may not be capable of determining the
+        /// number of ereports that were lost, or even *if* data loss actually
+        /// occurred. Therefore, a `None` here indicates *possible* data loss,
+        /// while a `Some(u32)` indicates *known* data loss.
+        lost: Option<u32>,
```

Comment:
Nit: I understand the desire to reuse an `Option` here. However, based on the english language meanings it may be misleading to see `lost: None`. It may be better to use a custom enum like:

```rust
enum Lost {
    Unknown,
    Num(u32)
}
```

Reply:
Hmm, I went with an `Option` because I was hoping to use a JSON `null` to encode the "unknown" case rather than a whole new enum tag and such in the JSON schema. I think it's possible to cajole `serde` into serializing a user-defined enum as "number or `null`" though; I can look into that if you think it's worthwhile?

Reply:
I think a bunch of us have become disenchanted at `Option` because of how easy it is to overload it semantically in code.

I guess my question is, is it possible to semantically overload it in serialized state too? If so we may want to use a user-defined enum everywhere.

Reply:
How do you feel about https://github.com/oxidecomputer/omicron/pull/7833/commits/3c5eb138f41aa44b1f67c38260ce380c5b57dd31 ?